<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CosMx Workflow • BBCC.OSTA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CosMx Workflow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BBCC.OSTA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/img-workflow-cosmx.html">CosMx Workflow</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drighelli/BBCC_OSTA" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CosMx Workflow</h1>
                        <h4 data-toc-skip class="author">OSTA Group</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/drighelli/BBCC_OSTA/blob/master/vignettes/img-workflow-cosmx.Rmd" class="external-link"><code>vignettes/img-workflow-cosmx.Rmd</code></a></small>
      <div class="hidden name"><code>img-workflow-cosmx.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="workflow-cosmx-data-analysis">Workflow: CosMX data analysis<a class="anchor" aria-label="anchor" href="#workflow-cosmx-data-analysis"></a>
</h2>
<p>This workflow is part of a series of vignettes illustrating analysis
workflows for spatial transcriptomics data using the <em><a href="https://bioconductor.org/packages/3.22/OSTA" class="external-link">OSTA</a></em>
ecosystem. OSTA is a textbook aimed at providing practical guidance for
the analysis of spatial transcriptomics data using Bioconductor
packages. The book is available at <a href="https://bioconductor.org/books/release/OSTA/" class="external-link">this link</a>, a
preprint of the book is available at <a href="https://www.biorxiv.org/content/10.1101/2025.11.20.688607v1" class="external-link">this
link</a>, and the code repository is at <a href="https://github.com/lmweber/OSTA" class="external-link">this link</a>.</p>
<p>This workflow extends the existing workflow present in OSTA book,
with additional details specifically suited for the <a href="https://www.bbcc-meetings.it/" class="external-link">BBCC25 conference</a>.</p>
<div class="section level3">
<h3 id="instructors">Instructors:<a class="anchor" aria-label="anchor" href="#instructors"></a>
</h3>
<ul>
<li>Dario Righelli<sup>1</sup>
</li>
<li>Francesca Calanca<sup>2,3</sup>
</li>
<li>Emanuela Scafuri<sup>2,3</sup>
</li>
</ul>
<p>1 Department of Electrical Engineer and Information Technology,
University of Naples “Federico II”, Italy<br>
2 Department of Chemical, Materials and Production Engineering,
University of Naples “Federico II”, Italy<br>
3 Institute for Applied Mathematics “M. Picone”, IAC-CNR, Naples,
Italy</p>
</div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this demo, we will analyze a 1k-plex CosMx data (Bruker) of a
mouse brain sample, see other datasets at <a href="https://bioconductor.org/books/release/OSTA/pages/bkg-example-datasets.html" class="external-link">this
link</a> for additional examples and exercising yourself. Main shown
steps are quality control and preprocessing, cell type annotation,
identification of marker genes, and cell-cell interaction analysis.</p>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<p>We now load all the required libraries.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/peicai/DESpace" class="external-link">DESpace</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/estellad/OSTA.data" class="external-link">OSTA.data</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scrapper</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpaceTrooper" class="external-link">SpaceTrooper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/estellad/SpatialExperimentIO" class="external-link">SpatialExperimentIO</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mjemons/spatialFDA" class="external-link">spatialFDA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="co"># set seed for random number generation</span></span>
<span><span class="co"># in order to make results reproducible</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">112358</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<div class="section level3">
<h3 id="cosmx-data">CosMx data<a class="anchor" aria-label="anchor" href="#cosmx-data"></a>
</h3>
<p>We load the dataset which was pre-stored in the
<code>OSTA.data</code> package. The dataset contains about 50,000 cells
profiled for about 1,000 RNA targets. We then use the
<code>readCosmxSXE</code> function from the
<code>SpatialExperimentIO</code> package to read the data into a
<code>SpatialExperiment</code> object <span class="citation">Righelli et
al. (<a href="#ref-Righelli2022-SpatialExperiment">2022</a>)</span>.
Then, it is possible to adapt the dataset to work with
<code>SpaceTrooper</code> by updating the object and reading the cell
polygons (this process is optional if we don’t need the polygons).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve CosMx dataset from OSF repo</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="st">"CosMx1k_MouseBrain2"</span></span>
<span><span class="va">pa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/OSTA.data/man/OSTA.data.html" class="external-link">OSTA.data_load</a></span><span class="op">(</span><span class="va">id</span>, mol<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">pa</span>, exdir<span class="op">=</span><span class="va">td</span><span class="op">)</span></span>
<span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperimentIO/man/readCosmxSXE.html" class="external-link">readCosmxSXE</a></span><span class="op">(</span><span class="va">td</span>, addTx<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># prepare data for 'SpaceTrooper'</span></span>
<span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/updateCosmxSPE.html" class="external-link">updateCosmxSPE</a></span><span class="op">(</span><span class="va">cos</span>, <span class="va">td</span>, sampleName<span class="op">=</span><span class="st">"CosMx"</span><span class="op">)</span></span>
<span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/readAndAddPolygonsToSPE.html" class="external-link">readAndAddPolygonsToSPE</a></span><span class="op">(</span><span class="va">cos</span><span class="op">)</span></span>
<span><span class="va">cos</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span> <span class="co"># artifact to work with ggspavis</span></span>
<span><span class="va">cos</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 950 48556 </span></span>
<span><span class="co">## metadata(4): fov_positions fov_dim polygons technology</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(950): Chrna4 Slc6a1 ... Cck Aqp4</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(48556): f1_c1 f1_c10 ... f99_c98 f99_c99</span></span>
<span><span class="co">## colData names(22): fov cellID ... polygons in_tissue</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(1): NegPrb</span></span>
<span><span class="co">## spatialCoords names(2) : CenterX_global_px CenterY_global_px</span></span>
<span><span class="co">## imgData names(1): sample_id</span></span></code></pre>
<p>In alternative, it is possible to directly load imaging-based
technologies dataset using the SpaceTrooper package with one of the
following commands. Note that for adding the poligons to the
SpatialExperiment object, you can use the
<code>readAndAddPolygonsToSPE</code> function as shown above, or just
use the keepPolygons=TRUE argument (for xenium and merfish) in the
reading functions.</p>
<p>NB: these are only for example purpose. If you want to download
xenium or merfish data, you need to find a dataset of one of these
technologies (i.e. in OSTA.data) and then put the path of the dataset to
the functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/readCosmxSPE.html" class="external-link">readCosmxSPE</a></span><span class="op">(</span><span class="va">td</span>, sampleName <span class="op">=</span> <span class="st">"CosMx"</span><span class="op">)</span></span>
<span><span class="va">cospro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/readCosmxSPE.html" class="external-link">readCosmxProteinSPE</a></span><span class="op">(</span><span class="va">td</span>, sampleName <span class="op">=</span> <span class="st">"CosMxProtein"</span><span class="op">)</span></span>
<span><span class="va">xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/readXeniumSPE.html" class="external-link">readXeniumSPE</a></span><span class="op">(</span>dirName <span class="op">=</span> <span class="st">"path/to/xenium/data"</span>, sampleName <span class="op">=</span> <span class="st">"Xenium"</span>, keepPolygons <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># do not run as it is</span></span>
<span><span class="va">mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/readMerfishSPE.html" class="external-link">readMerfishSPE</a></span><span class="op">(</span>dirName <span class="op">=</span> <span class="st">"path/to/merscope/data"</span>, sampleName <span class="op">=</span> <span class="st">"MERSCOPE"</span>, keepPolygons <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># do not run as it is</span></span></code></pre></div>
<p>Once loaded, you can access additional metadata of the
SpatialExperiment object <code>cos</code> as follows: - colData(cos):
cell-level metadata, including FOV ID, cell area, and QC metrics. -
alternatively, you can directly access a specific metadata column with
<code>cos$&lt;column_name&gt;</code>. (e.g., <code>cos$fov</code> gives
the FOV IDs for each cell). - rowData(cos): gene-level metadata,
including gene names and biotypes. - assays(cos): names of the assays
available in the object (e.g., “counts” for raw counts, “logcounts” for
log-normalized counts if available). - spatialCoords(cos): spatial
coordinates of the cells.</p>
</div>
</div>
<div class="section level2">
<h2 id="basic-visualization">Basic Visualization<a class="anchor" aria-label="anchor" href="#basic-visualization"></a>
</h2>
<p>We can have an overview of the data, by plotting the spatial
distribution of the cells, represented by their centroids. This is
useful to have a global bird’s-eye-view of the tissue.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotCentroids.html" class="external-link">plotCentroids</a></span><span class="op">(</span><span class="va">cos</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/plot-xy-1.png" class="r-plt" width="700"></p>
<p>Unlike, Xenium and MERSCOPE, CosMx does not stitch fields of view
(FOVs) in a single image, but rather treats each FOV independently. It
is therefore useful to visualize a map of the FOVs to understand their
spatial arrangement.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotCellsFovs.html" class="external-link">plotCellsFovs</a></span><span class="op">(</span><span class="va">cos</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/plot-fov-map-1.png" class="r-plt" width="66%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="excercises">Excercises:<a class="anchor" aria-label="anchor" href="#excercises"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Explore the <code>cos</code> object using <code>colData(cos)</code>,
<code>rowData(cos)</code>, and <code>assays(cos)</code>.</li>
<li>How many cells and genes are in the dataset? (hint: use
<code>dim(cos)</code>)</li>
<li>Explore the cell metadata. What information is available for each
cell? Do you recognize something specific to CosMx data?</li>
<li>Check the distribution of counts across cells. Are there any cells
with very low counts?</li>
<li>Extract the spatial coordinates of the cells using
<code>spatialCoords(cos)</code>. What is the range of the x and y
coordinates?</li>
<li>plot the centroids of the cell using
<code>plotCentroids(cos)</code>. Colouring them for a specific cell
metadata.</li>
<li>
<strong>BONUS</strong>: Load a different Xenium dataset from the
<code>OSTA.data</code> package for further analysis (use the
<code>Janesick_breastCancer_Xenium_rep1</code>). This could be heavy to
load for your laptop if you think to do it together with the cosmx
dataset.</li>
</ol>
</div>
<div class="section level2">
<h2 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p>Quality control (QC) is a crucial step in imaging-based spatial
transcriptomics, just as it is in scRNA-seq, where it helps remove
low-quality cells and reveal technical or biological biases at the
sample, field-of-view (FOV), or slide-area level.</p>
<p>A key difference, however, is that imaging-based ST platforms
typically do not quantify the whole transcriptome—or even an unbiased
subset of it—but rely instead on predefined gene panels. These panels,
depending on platform and experimental goals, may be pan-tissue or
focused on specific tissues or cell types. Their size varies widely,
from a few hundred genes (e.g., 313 genes in the first Xenium panel) to
several thousand (e.g., the Xenium 5k panel or the CosMx 6k and 18k
panels). As a result, unlike sequencing-based assays, imaging platforms
may not provide enough ribosomal or mitochondrial transcripts to compute
standard QC metrics. They do, however, include sets of negative probes
that allow detection of non-specific RNA capture.</p>
<p>Finally, imaging-based ST technologies also offer rich morphological
information (e.g., cell area, eccentricity) and signals from antibody
stains—such as nuclear or membrane markers—which can further support QC
and downstream analyses.</p>
<p>Here we compute the cell-level QC metrics using the
<code>SpaceTrooper</code> package.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">old</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">cos</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lys</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NegPrb"</span>, <span class="st">"Negative"</span>, <span class="st">"SystemControl"</span><span class="op">)</span></span>
<span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/spatialPerCellQC.html" class="external-link">spatialPerCellQC</a></span><span class="op">(</span><span class="va">cos</span>, rmZeros<span class="op">=</span><span class="cn">TRUE</span>, negProbList<span class="op">=</span><span class="va">lys</span><span class="op">)</span></span></code></pre></div>
<p>This produces the following new QC metrics in the object’s
<code>colData</code>:</p>
<pre><code><span><span class="co">##  [1] "sum"                     "detected"               </span></span>
<span><span class="co">##  [3] "altexps_NegPrb_sum"      "altexps_NegPrb_detected"</span></span>
<span><span class="co">##  [5] "altexps_NegPrb_percent"  "total"                  </span></span>
<span><span class="co">##  [7] "control_sum"             "control_detected"       </span></span>
<span><span class="co">##  [9] "target_sum"              "target_detected"        </span></span>
<span><span class="co">## [11] "CenterX_global_px"       "CenterY_global_px"      </span></span>
<span><span class="co">## [13] "ctrl_total_ratio"        "log2Ctrl_total_ratio"   </span></span>
<span><span class="co">## [15] "CenterX_global_um"       "CenterY_global_um"      </span></span>
<span><span class="co">## [17] "Area_um"                 "dist_border_x"          </span></span>
<span><span class="co">## [19] "dist_border_y"           "dist_border"            </span></span>
<span><span class="co">## [21] "log2AspectRatio"         "CountArea"              </span></span>
<span><span class="co">## [23] "log2CountArea"</span></span></code></pre>
<p>If we want to explore the distribution of a specific QC metric, e.g.,
the log2 of counts per area, we can use the <code>plotMetricHist</code>
function. When computing the spatial outliers, we can use either the
scuttle (sc) or medcaple (mc) methods, or both. Then we can visualize
the two outlier thresholds on the histogram.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/computeSpatialOutlier.html" class="external-link">computeSpatialOutlier</a></span><span class="op">(</span><span class="va">cos</span>, computeBy <span class="op">=</span> <span class="st">"log2CountArea"</span>, method<span class="op">=</span><span class="st">"both"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotMetricHist.html" class="external-link">plotMetricHist</a></span><span class="op">(</span><span class="va">cos</span>, <span class="st">"log2CountArea"</span>, useFences <span class="op">=</span> <span class="st">"log2CountArea_outlier_sc"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/unnamed-chunk-3-1.png" class="r-plt" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotMetricHist.html" class="external-link">plotMetricHist</a></span><span class="op">(</span><span class="va">cos</span>, <span class="st">"log2CountArea"</span>, useFences <span class="op">=</span> <span class="st">"log2CountArea_outlier_mc"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/unnamed-chunk-3-2.png" class="r-plt" width="700"></p>
<p>Or visualize the metric in the spatial context:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotCentroids.html" class="external-link">plotCentroids</a></span><span class="op">(</span><span class="va">cos</span>, colourBy<span class="op">=</span><span class="st">"log2CountArea"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coordinate system already present.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Adding new coordinate system, which will replace the existing one.</span></span></code></pre>
<p><img src="img-workflow-cosmx_files/figure-html/unnamed-chunk-4-1.png" class="r-plt" width="700"></p>
<p>Now, we will simply compute <em><a href="https://bioconductor.org/packages/3.22/SpaceTrooper" class="external-link">SpaceTrooper</a></em>’s
aggregated QC score and keep the cells that pass the threshold. More
details will be provided once the paper will be published (Banzi ,
Righelli et al., in preparation).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/computeQCScore.html" class="external-link">computeQCScore</a></span><span class="op">(</span><span class="va">cos</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Metric log2Ctrl_total_ratio produced NA skewness values.It will be removed from the QC score formula.</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotCentroids.html" class="external-link">plotCentroids</a></span><span class="op">(</span><span class="va">cos</span>, colourBy<span class="op">=</span><span class="st">"QC_score"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coordinate system already present.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Adding new coordinate system, which will replace the existing one.</span></span></code></pre>
<p><img src="img-workflow-cosmx_files/figure-html/unnamed-chunk-5-1.png" class="r-plt" width="700"></p>
<p>After that we can compute the QC flags, visualize them and filter out
low-quality cells.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/computeQCScoreFlags.html" class="external-link">computeQCScoreFlags</a></span><span class="op">(</span><span class="va">cos</span>, qsThreshold<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotCentroids.html" class="external-link">plotCentroids</a></span><span class="op">(</span><span class="va">cos</span>, colourBy<span class="op">=</span><span class="st">"low_qcscore"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.key.size<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>col<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/filter-qc-cosmx-1.png" class="r-plt" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpaceTrooper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotMetricHist.html" class="external-link">plotMetricHist</a></span><span class="op">(</span><span class="va">cos</span>, <span class="st">"QC_score"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cos</span> <span class="op">&lt;-</span> <span class="va">cos</span><span class="op">[</span>, <span class="op">!</span><span class="va">cos</span><span class="op">$</span><span class="va">low_qcscore</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="excercises-1">Excercises:<a class="anchor" aria-label="anchor" href="#excercises-1"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Explore the newly added QC metrics in <code>colData(cos)</code>.
What do they represent?</li>
<li>Visualize the distribution of another QC metric, e.g.,
`log2NegativeArea</li>
<li>Are you able to identify outlier cells based on this metric? how you
visualize them?</li>
<li>Visualize the spatial distribution of another QC metric.</li>
<li>Compute the outliers for the QC_score (before filtering low qc
cells) using both methods (sc and mc). How many cells are flagged as low
quality by each method?</li>
</ol>
</div>
<div class="section level2">
<h2 id="processing">Processing<a class="anchor" aria-label="anchor" href="#processing"></a>
</h2>
<p>For the sake of runtime and laptop phisical limitations, we will
perform downstream analyses only on a small portion of the tissue,
defined by the following FOVs:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">72</span><span class="op">:</span><span class="fl">74</span>, <span class="fl">90</span><span class="op">:</span><span class="fl">92</span>, <span class="fl">97</span><span class="op">:</span><span class="fl">99</span>, <span class="fl">114</span><span class="op">:</span><span class="fl">116</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotZoomFovsMap.html" class="external-link">plotZoomFovsMap</a></span><span class="op">(</span><span class="va">cos</span>, fovs<span class="op">=</span><span class="va">fs</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/subset-fovs-cosmx-1.png" class="r-plt" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sub</span> <span class="op">&lt;-</span> <span class="va">cos</span><span class="op">[</span>, <span class="va">cos</span><span class="op">$</span><span class="va">fov</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">fs</span><span class="op">]</span></span></code></pre></div>
<p>Next, we’ll log-normalize counts by area, and perform principal
component analysis (PCA) on all 950 RNA targets:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cell area-based normalization</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span>, <span class="fl">2</span>, <span class="va">sub</span><span class="op">$</span><span class="va">Area</span>, <span class="va">`/`</span><span class="op">)</span></span>
<span><span class="co"># principal component analysis</span></span>
<span><span class="va">sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span></span></code></pre></div>
<p>Let’s visualize the expression of the first two principal components
(PCs) in the spatial context.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add PCs as cell metadata</span></span>
<span><span class="va">pcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sub</span>, <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span>, <span class="va">pcs</span><span class="op">)</span></span>
<span><span class="co"># visualize PCs 1 &amp; 2 in space</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotCentroids.html" class="external-link">plotCentroids</a></span><span class="op">(</span><span class="va">sub</span>, colourBy<span class="op">=</span><span class="st">"PC1"</span><span class="op">)</span> <span class="op">+</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotCentroids.html" class="external-link">plotCentroids</a></span><span class="op">(</span><span class="va">sub</span>, colourBy<span class="op">=</span><span class="st">"PC2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/plt-gs-xy-1.png" class="r-plt" width="768"></p>
</div>
<div class="section level2">
<h2 id="cell-type-annotation">Cell type Annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h2>
<p>Cell type annotation is a key step in the analysis of spatial
transcriptomics data, as it allows us to assign biological meaning to
the observed spatial patterns of gene expression.</p>
<p>Here, we use the <span class="citation"><span class="nocase">Zeisel
et al.</span> (<a href="#ref-zeisel2015cell">2015</a>)</span> dataset as
a reference to annotate the cell types with <em><a href="https://bioconductor.org/packages/3.22/SingleR" class="external-link">SingleR</a></em>.
This dataset is composed by single-cell RNA-seq data from the mouse
brain, and contains 3000 cells from a variety of annotated cell types at
two different levels of granularity (level 1 and level 2). We will use
the coarse level 1 annotations for this demo.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html" class="external-link">ZeiselBrainData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sceZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sceZ</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sceZ</span><span class="op">$</span><span class="va">level1class</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## astrocytes_ependymal    endothelial-mural         interneurons </span></span>
<span><span class="co">##                  224                  235                  290 </span></span>
<span><span class="co">##            microglia     oligodendrocytes        pyramidal CA1 </span></span>
<span><span class="co">##                   98                  820                  939 </span></span>
<span><span class="co">##         pyramidal SS </span></span>
<span><span class="co">##                  399</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sceZ</span><span class="op">$</span><span class="va">level2class</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    (none)    Astro1    Astro2   CA1Pyr1   CA1Pyr2 CA1PyrInt   CA2Pyr2   Choroid </span></span>
<span><span class="co">##       189        68        61       380       447        49        41        10 </span></span>
<span><span class="co">##   ClauPyr     Epend      Int1     Int10     Int11     Int12     Int13     Int14 </span></span>
<span><span class="co">##         5        20        12        21        10        21        15        22 </span></span>
<span><span class="co">##     Int15     Int16      Int2      Int3      Int4      Int5      Int6      Int7 </span></span>
<span><span class="co">##        18        20        24        10        15        20        22        23 </span></span>
<span><span class="co">##      Int8      Int9      Mgl1      Mgl2    Oligo1    Oligo2    Oligo3    Oligo4 </span></span>
<span><span class="co">##        26        11        17        16        45        98        87       106 </span></span>
<span><span class="co">##    Oligo5    Oligo6     Peric      Pvm1      Pvm2   S1PyrDL  S1PyrL23   S1PyrL4 </span></span>
<span><span class="co">##       125       359        21        32        33        81        74        26 </span></span>
<span><span class="co">##   S1PyrL5  S1PyrL5a   S1PyrL6  S1PyrL6b    SubPyr     Vend1     Vend2      Vsmc </span></span>
<span><span class="co">##        16        28        39        21        22        32       105        62</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span>test<span class="op">=</span><span class="va">sub</span>, ref<span class="op">=</span><span class="va">sceZ</span>, labels<span class="op">=</span><span class="va">sceZ</span><span class="op">$</span><span class="va">level1class</span>, de.method<span class="op">=</span><span class="st">"wilcox"</span><span class="op">)</span></span>
<span><span class="va">sub</span><span class="op">$</span><span class="va">SingleR_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">pruned.labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nlevels.html" class="external-link">nlevels</a></span><span class="op">(</span><span class="va">sub</span><span class="op">$</span><span class="va">SingleR_label</span><span class="op">)</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="va">nk</span>, <span class="st">"Spectral"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotPolygons.html" class="external-link">plotPolygons</a></span><span class="op">(</span><span class="va">sub</span>, </span>
<span>    colourBy<span class="op">=</span><span class="st">"SingleR_label"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/annot-cosmx-1.png" class="r-plt" width="700"></p>
<p>Note that the annotation performed here is a transfer label from a
reference dataset that only partially matches with the tissue under
study and purposely annotated at a coarse level (level 1). Ideally, one
can use a more appropriate reference dataset, or better yet, build a
custom reference from single-cell RNA-seq data of the same tissue.</p>
<p>A different approach consists in using, in addition to the gene
expression profiles, the image data and the spatial context of the
cells. <em><a href="https://github.com/Nanostring-Biostats/InSituType" class="external-link">InSituType</a></em>
<span class="citation">(<a href="#ref-danaher2022insitutype"><span class="nocase">Danaher et al.</span> 2022</a>)</span> provides a
framework to perform such analysis in unsupervised, semi-supervised, or
supervised manners. Please refer to its documentation for more
details.</p>
</div>
<div class="section level2">
<h2 id="marker-genes">Marker genes<a class="anchor" aria-label="anchor" href="#marker-genes"></a>
</h2>
<p>Once that we have identified the cell types, we can identify
cluster-related spatially-variable genes using the <em><a href="https://bioconductor.org/packages/3.22/DESpace" class="external-link">DESpace</a></em>
package <span class="citation">(<a href="#ref-Cai2024-DESpace">Cai et
al. 2024</a>)</span>. In this workflow, we use this as a way to identify
marker genes for the cell types annotated above.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://peicai.github.io/DESpace/reference/svg_test.html" class="external-link">svg_test</a></span><span class="op">(</span>spe<span class="op">=</span><span class="va">sub</span>, cluster_col<span class="op">=</span><span class="st">"SingleR_label"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">gene_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         gene_id       LR   logCPM PValue FDR</span></span>
<span><span class="co">## Itm2a     Itm2a 2596.761 11.03859      0   0</span></span>
<span><span class="co">## Flt1       Flt1 2601.543 11.05386      0   0</span></span>
<span><span class="co">## Cldn5     Cldn5 2825.269 11.11885      0   0</span></span>
<span><span class="co">## Gja1       Gja1 4982.193 11.29054      0   0</span></span>
<span><span class="co">## Gpr37l1 Gpr37l1 4989.928 11.30344      0   0</span></span>
<span><span class="co">## Bcan       Bcan 3567.078 11.61163      0   0</span></span></code></pre>
<p>We focus on the genes with the smallest p-values, and compute the
average expression of each gene in each cell type. We can visualize this
with a heatmap.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">gene_results</span></span>
<span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/rank.html" class="external-link">rank</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">PValue</span>, ties.method<span class="op">=</span><span class="st">"min"</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">mu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span><span class="op">[</span><span class="va">gs</span>, <span class="op">]</span>, <span class="fl">1</span>, <span class="va">tapply</span>, <span class="va">sub</span><span class="op">$</span><span class="va">SingleR_label</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">mu</span><span class="op">)</span>, </span>
<span>    scale<span class="op">=</span><span class="st">"column"</span>, show_colnames<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    cluster_rows<span class="op">=</span><span class="cn">TRUE</span>, cluster_cols<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>    main<span class="op">=</span><span class="st">"Average expression of top SVGs per cell type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/plt-svg-1.png" class="r-plt" width="768"></p>
<p>We can also visualize the spatial distribution of some of these
genes, e.g., Mbp that marks oligodendrocytes, and Calm1, Calm2, and
Snap25 that mark different subsets of neurons. Note also how some genes
show a difference of expression within the pyramidal neuron population,
suggesting the presence of subtypes.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mbp"</span>, <span class="st">"Calm1"</span>, <span class="st">"Calm2"</span>, <span class="st">"Snap25"</span><span class="op">)</span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">gs</span>, \<span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sub</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span><span class="op">[</span><span class="va">g</span>, <span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpaceTrooper/man/plotPolygons.html" class="external-link">plotPolygons</a></span><span class="op">(</span><span class="va">sub</span>, colourBy<span class="op">=</span><span class="va">g</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> </span>
<span><span class="va">dy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sub</span><span class="op">)</span><span class="op">[</span><span class="va">gs</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">ps</span>, guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span></span>
<span>        <span class="st">"expression"</span>, </span>
<span>        limits<span class="op">=</span><span class="va">dy</span>, breaks<span class="op">=</span><span class="va">dy</span>,</span>
<span>        labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"low"</span>, <span class="st">"high"</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/plt-svg-genes-1.png" class="r-plt" width="960"></p>
</div>
<div class="section level2">
<h2 id="exercise">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Extract a list of marker genes and plot them</li>
</ol>
</div>
<div class="section level2">
<h2 id="neighborhood-based-gene-expression-analyses">Neighborhood-based gene expression analyses<a class="anchor" aria-label="anchor" href="#neighborhood-based-gene-expression-analyses"></a>
</h2>
<p>We start by creating a cell neighborhood graph, based on a distance
threshold of 50 microns. To do so, we use the <em><a href="https://bioconductor.org/packages/3.22/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></em>
package <span class="citation">(<a href="#ref-Moses2023-Voyager">Moses
et al. 2023</a>)</span>.</p>
<p>Note that an alternative approach uses the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
nearest neighbors (kNN) method to define the neighborhood graph. Here,
we prefer to use a distance-based approach, as it better captures the
morphological structure of this tissue area.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert to SFE &amp; remove cells with NA labels</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">sub</span>, <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">SingleR_label</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># this is needed for Voyager's plotting functions</span></span>
<span><span class="co"># sfe$polygons&lt;-.renameGeometry(sfe$polygons, from="global", to="geomery")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">polygons</span><span class="op">)</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"geometry"</span></span>
<span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">polygons</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"geometry"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cellSeg"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">polygons</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"poly2nb"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>,</span>
<span>        type<span class="op">=</span><span class="st">"cellSeg"</span>, </span>
<span>        method<span class="op">=</span><span class="st">"poly2nb"</span>, </span>
<span>        style<span class="op">=</span><span class="st">"W"</span>, snap<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph</a></span><span class="op">(</span><span class="va">sfe</span>, </span>
<span>    colGraphName<span class="op">=</span><span class="st">"poly2nb"</span>, </span>
<span>    colGeometryName<span class="op">=</span><span class="st">"cellSeg"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/cci-graph-cosmx-1.png" class="r-plt" width="700"></p>
<p>Given the neighborhood graph, we can calculate local measures of
spatial autocorrelation for specific genes in the dataset.</p>
<p>Moran’s I is a measure of spatial autocorrelation that quantifies the
degree to which observed values (e.g., gene expression) are similar
across neighboring spatial locations. The core idea is to compare the
value of each cell with the average value of its spatial neighbors,
using a spatial-weights matrix that defines what “neighboring”
means.</p>
<p>When Moran’s I is:</p>
<ul>
<li><p>positive → similar values cluster together (High–High or Low–Low
regions)</p></li>
<li><p>negative → dissimilar values tend to be adjacent Low-High
(checkerboard-likeLow patterns)</p></li>
<li><p>≈ 0 → no spatial structure (random pattern)</p></li>
</ul>
<p>Here, we focus on Snap25 and Calm1, neuronal marker genes identified
above that show interesting spatial expression distributions (see
above). First, we compute the local Moran’s I coefficient.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Snap25"</span>, <span class="st">"Calm1"</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>,</span>
<span>    features<span class="op">=</span><span class="va">gs</span>,</span>
<span>    type<span class="op">=</span><span class="st">"localmoran"</span>,</span>
<span>    zero.policy<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    colGraphName<span class="op">=</span><span class="st">"poly2nb"</span>,</span>
<span>    colGeometryName<span class="op">=</span><span class="st">"cellSeg"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, </span>
<span>    name<span class="op">=</span><span class="st">"localmoran"</span>,</span>
<span>    features<span class="op">=</span><span class="va">gs</span>, ncol<span class="op">=</span><span class="fl">2</span>,</span>
<span>    colGeometryName<span class="op">=</span><span class="st">"cellSeg"</span>,</span>
<span>    divergent<span class="op">=</span><span class="cn">TRUE</span>, diverge_center<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/cci-moran-snap25-1.png" class="r-plt" width="768"></p>
<p>From the plots, we can see that Snap25 shows positive spatial
autocorrelation in a very localized part of the tissue, while Calm1
shows a more diffuse autocorrelation structure.</p>
<p>We focus on Snap25 and visualize its autocorrelation structure using
a Moran scatter plot, which shows the relationship between the
expression of the gene in each cell and the average expression in its
neighbors.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>,</span>
<span>    feature<span class="op">=</span><span class="va">gs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    zero.policy<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    type<span class="op">=</span><span class="st">"moran.plot"</span>,</span>
<span>    colGraphName<span class="op">=</span><span class="st">"poly2nb"</span>,</span>
<span>    colGeometryName<span class="op">=</span><span class="st">"cellSeg"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/moranPlot.html" class="external-link">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>,</span>
<span>    feature<span class="op">=</span><span class="va">gs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    graphName<span class="op">=</span><span class="st">"poly2nb"</span>,</span>
<span>    swap_rownames<span class="op">=</span><span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/cci-moran-scatter-snap25-1.png" class="r-plt" width="700"></p>
<p>We can then classify cells based on the significance of their local
Moran’s I values, and visualize the resulting clusters in the spatial
context.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResults</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="st">"localmoran"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">locClust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>    <span class="va">`-log10p_adj`</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span>, <span class="st">"non-siginificant"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResults</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="st">"localmoran"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">res</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>,</span>
<span>    features<span class="op">=</span><span class="va">gs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    name<span class="op">=</span><span class="st">"localmoran"</span>,</span>
<span>    attribute<span class="op">=</span><span class="st">"locClust"</span>,</span>
<span>    colGeometryName<span class="op">=</span><span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/cci-moran-scatter-sig-1.png" class="r-plt" width="700"></p>
<p>In this case, we can see a significant positive autocorrelation in
the lower left part of the structure, while the rest of the tissue does
not show significant autocorrelation.</p>
<p>Note that other local spatial autocorrelation measures, such as
Geary’s C coefficient and Getis-Ord statistic, as well as their global
versions can be used for similar analyses <span class="citation">Moses
et al. (<a href="#ref-Moses2023-Voyager">2023</a>)</span>.</p>
</div>
<div class="section level2">
<h2 id="cell-cell-interactions">Cell-cell interactions<a class="anchor" aria-label="anchor" href="#cell-cell-interactions"></a>
</h2>
<div class="section level3">
<h3 id="joint-counts">Joint counts<a class="anchor" aria-label="anchor" href="#joint-counts"></a>
</h3>
<p>We can also use the neighborhood graph to investigate cell-cell
interactions between different cell types annotated above. Here, we use
the join count statistic implemented in the <em><a href="https://CRAN.R-project.org/package=spdep" class="external-link">spdep</a></em> package
to quantify the tendency of cells of different types to be
neighbors.</p>
<p>In practice, join counts evaluate, for each pair of labels, how
frequently two adjacent cells belong to those categories, and compare
this to an expectation derived from a permutation-based null model. The
resulting z-scores indicate the strength and direction of the
interaction:</p>
<ul>
<li><p>Positive z-values → the two cell types co-locate more than
expected (attraction / spatial clustering).</p></li>
<li><p>Negative z-values → the two cell types co-locate less than
expected (avoidance / spatial segregation).</p></li>
<li><p>Values near zero → no significant deviation from spatial
randomness.</p></li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">jc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/spdep/reference/joincount.multi.html" class="external-link">joincount.multi</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">SingleR_label</span>, <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"poly2nb"</span><span class="op">)</span>, zero.policy<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">jc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">jc</span><span class="op">[</span>, <span class="st">"z-value"</span><span class="op">]</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                     Joincount   Expected  Variance   z-value</span></span>
<span><span class="co">## Jtot                               1132.15707 2494.52543 265.87369 -83.55200</span></span>
<span><span class="co">## pyramidal CA1:pyramidal CA1         735.66944  351.21205  62.92504  48.46592</span></span>
<span><span class="co">## pyramidal CA1:oligodendrocytes       22.39901  303.28546  75.62290 -32.30012</span></span>
<span><span class="co">## oligodendrocytes:oligodendrocytes   196.71667   65.35907  19.34973  29.86192</span></span>
<span><span class="co">## pyramidal CA1:astrocytes_ependymal   97.55301  379.29404  92.67327 -29.26664</span></span>
<span><span class="co">## pyramidal SS:pyramidal CA1          195.38757  445.19310 106.83889 -24.16781</span></span></code></pre>
<p>We can see that there is a strong tendency for cells of the same type
to be neighbors (high positive z-values). There are also some
interesting negative interactions between different cell types: e.g.,
pyramidal CA1 tend to be away from oligodendrocytes and astrocytes. This
highlights the compartimentalized structure of the brain tissue.</p>
</div>
<div class="section level3">
<h3 id="point-pattern-analysis">Point-pattern analysis<a class="anchor" aria-label="anchor" href="#point-pattern-analysis"></a>
</h3>
<p>We complement the graph-based interaction analysis with point-pattern
statistics, which treat each cell as a spatial point and quantify
deviations from complete spatial randomness (CSR). Using the <em><a href="https://bioconductor.org/packages/3.22/spatialFDA" class="external-link">spatialFDA</a></em>
package, we compute Besag’s L function, a variance-stabilized form of
Ripley’s K function, to characterize both within-type and between-type
spatial interactions.</p>
<p>Besag’s <code>L(r)</code> summarizes how many neighboring cells occur
within distance <code>r</code> of a reference cell, compared to the
expectation under CSR. When the empirical curve lies above the CSR line,
the point pattern exhibits clustering, whereas values below the line
indicate dispersion or spatial inhibition.</p>
<p>Keep in mind that Ripley’s K function, and the pair correlation
function <span class="citation">(<a href="#ref-Emons2025-pasta">Emons et
al. 2025</a>)</span> makes the quite strong assuption of homogeneity of
the point pattern, which is irrealistic in tissue biology. In
particular, CSR is a naive null model that does not take into account
tissue structure and cell density variations across the tissue. Hence,
these plots should be considered as exploratory analyses rather than
formal hypothesis tests.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spatialFDA/man/calcMetricPerFov.html" class="external-link">calcMetricPerFov</a></span><span class="op">(</span></span>
<span>    spe<span class="op">=</span><span class="va">sub</span>, fun<span class="op">=</span><span class="st">"Lest"</span>,</span>
<span>    subsetby<span class="op">=</span><span class="st">"sample_id"</span>, </span>
<span>    marks<span class="op">=</span><span class="st">"SingleR_label"</span>,</span>
<span>    selection<span class="op">=</span><span class="st">"pyramidal CA1"</span>,</span>
<span>    rSeq<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">500</span>, l<span class="op">=</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    by<span class="op">=</span><span class="st">"sample_id"</span>, ncores<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/spatialFDA/man/plotMetricPerFov.html" class="external-link">plotMetricPerFov</a></span><span class="op">(</span><span class="va">res</span>, </span>
<span>    theo<span class="op">=</span><span class="cn">TRUE</span>, correction<span class="op">=</span><span class="st">"iso"</span>, </span>
<span>    x<span class="op">=</span><span class="st">"r"</span>, imageId<span class="op">=</span><span class="st">"sample_id"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/cci-CA1-l-1.png" class="r-plt" width="576"></p>
<p>In this type of plots, the dashed line represents the theoretical
value of the metric under complete spatial randomness (CSR). Values
above this line indicate clustering, while values below indicate
dispersion. Here, we can see that pyramidal CA1 neurons tend to cluster
together, as expected.</p>
<p>Note that the L-function lays below the CSR line for the first 50
microns, which is a consequence of the cell segmentation that does not
allow cells to overlap, indicating the impossibility of finding two
cells at a distance smaller than their size.</p>
<p>While in the plot above we focused on a single cell type, we can also
investigate the interaction between two different cell types, e.g.,
pyramidal CA1 neurons and oligodendrocytes, using the cross-L
function.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spatialFDA/man/calcMetricPerFov.html" class="external-link">calcMetricPerFov</a></span><span class="op">(</span></span>
<span>    spe<span class="op">=</span><span class="va">sub</span>, fun<span class="op">=</span><span class="st">"Lcross"</span>, </span>
<span>    subsetby<span class="op">=</span><span class="st">"sample_id"</span>, </span>
<span>    marks<span class="op">=</span><span class="st">"SingleR_label"</span>,</span>
<span>    selection<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pyramidal CA1"</span>, <span class="st">"oligodendrocytes"</span><span class="op">)</span>,</span>
<span>    rSeq<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">500</span>, l<span class="op">=</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    by<span class="op">=</span><span class="st">"sample_id"</span>, ncores<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/spatialFDA/man/plotMetricPerFov.html" class="external-link">plotMetricPerFov</a></span><span class="op">(</span><span class="va">res</span>, </span>
<span>    theo<span class="op">=</span><span class="cn">TRUE</span>, correction<span class="op">=</span><span class="st">"iso"</span>, </span>
<span>    x<span class="op">=</span><span class="st">"r"</span>, imageId<span class="op">=</span><span class="st">"sample_id"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img-workflow-cosmx_files/figure-html/cci-cross-l-1.png" class="r-plt" width="576"></p>
<p>In this case, we can see that pyramidal CA1 neurons and
oligodendrocytes tend to be spatially segregated, as indicated by the
cross-L function being below the CSR line.</p>
<p>We only briefly discussed cell-cell interactions and spatial
exploratory analyses here: we refer interested readers to the
comprehensive documentation and tutorials available at the <a href="https://robinsonlabuzh.github.io/pasta/00-home.html" class="external-link">pasta</a> and
<a href="https://pachterlab.github.io/voyager/" class="external-link">Voyager</a>
websites.</p>
</div>
</div>
<div class="section level2">
<h2 id="excercises-2">Excercises<a class="anchor" aria-label="anchor" href="#excercises-2"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Preprocessing: what percentage of variance is explained by PC1 and
PC2? Explore the explained percentage of variance with a plot
(i.e. elbow plot, histogram).</li>
<li>Annotation: create a frequency table of the annotation column. Which
is the most abundant cell type in this tissue subset? Which is the
rarest?</li>
<li>Do you know any specific markers for microglia? Try to explore their
expression. (i.e. extract the latest column from the heatmap)</li>
<li>The neighborhood graph was built using a distance snap of 50
microns. Try re-running, changing the distance snap e.g., 80 and compare
the two settings.</li>
</ol>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Cai2024-DESpace" class="csl-entry">
Cai, Peiying, Mark D Robinson, and Simone Tiberi. 2024. <span>“DESpace:
Spatially Variable Gene Detection via Differential Expression Testing of
Spatial Clusters.”</span> <em>Bioinformatics</em> 40 (btae027, 2). <a href="https://doi.org/10.1093/bioinformatics/btae027" class="external-link">https://doi.org/10.1093/bioinformatics/btae027</a>.
</div>
<div id="ref-danaher2022insitutype" class="csl-entry">
<span class="nocase">Danaher, Patrick, Edward Zhao, Zhi Yang, et
al.</span> 2022. <span>“Insitutype: Likelihood-Based Cell Typing for
Single Cell Spatial Transcriptomics.”</span> <em>BioRxiv</em>, 2022–10.
</div>
<div id="ref-Emons2025-pasta" class="csl-entry">
Emons, Martin, Samuel Gunz, Helena L. Crowell, Izaskun Mallona, Reinhard
Furrer, and Mark D. Robinson. 2025. <span>“Harnessing the Potential of
Spatial Statistics for Spatial Omics Data with Pasta.”</span>
<em>Nucleic Acids Research</em> 53 (17): gkaf870. <a href="https://doi.org/10.1038/s41467-022-28020-5" class="external-link">https://doi.org/10.1038/s41467-022-28020-5</a>.
</div>
<div id="ref-Moses2023-Voyager" class="csl-entry">
Moses, Lambda, Pétur Helgi Einarsson, Kayla Jackson, et al. 2023.
<span>“Voyager: Exploratory Single-Cell Genomics Data Analysis with
Geospatial Statistics.”</span> <em>bioRxiv</em>, ahead of print. <a href="https://doi.org/10.1101/2023.07.20.549945" class="external-link">https://doi.org/10.1101/2023.07.20.549945</a>.
</div>
<div id="ref-Righelli2022-SpatialExperiment" class="csl-entry">
Righelli, Dario, Lukas M Weber, Helena L Crowell, et al. 2022.
<span>“SpatialExperiment: Infrastructure for Spatially-Resolved
Transcriptomics Data in r Using Bioconductor.”</span>
<em>Bioinformatics</em> 38: 3128–31. <a href="https://doi.org/10.1093/bioinformatics/btac299" class="external-link">https://doi.org/10.1093/bioinformatics/btac299</a>.
</div>
<div id="ref-zeisel2015cell" class="csl-entry">
<span class="nocase">Zeisel, Amit, Ana B Muñoz-Manchado, Simone
Codeluppi, et al.</span> 2015. <span>“Cell Types in the Mouse Cortex and
Hippocampus Revealed by Single-Cell RNA-Seq.”</span> <em>Science</em>
347 (6226): 1138–42.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Dario Righelli, Francesca Calanca, Emanuela Scafuri.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
